// d4-roshamresult/d4-roshamresult.js
var mongoose = require('mongoose');
var weaponTypes = ['Rock','Scissors','Paper'];
var schResult = new mongoose.Schema({
    turnDate: Date,
    player1: mongoose.Schema.Types.ObjectId, 
    weapon1: {type: String, enum: weaponTypes},
    player2: mongoose.Schema.Types.ObjectId, 
    weapon2: {type: String, enum: weaponTypes},
    winner: Number
});
var modResult = mongoose.model('result', schResult);


function whoWins(weapon1, weapon2){
    var weaponVal1 = weaponTypes.indexOf(weapon1);
    var weaponVal2 = weaponTypes.indexOf(weapon2);
    var battleSum = weaponVal1 + weaponVal2;
    
    if(weaponVal1 === weaponVal2) return 0;
    if(weaponVal1 > weaponVal2){
        if(battleSum === 2) return 1;
        else return 2;
    }
    else{
        if(battleSum === 2) return 2;
        else return 1;
    }
};

//************************************************************************************************************
// function     : d4-roshamresult.logResult
// developer    : Kai
// developed on : 5/19/2014
// arguments    : battle object {turnDate, player1, weapon1, player2, weapon2} - player 1 and 2 are userids
//************************************************************************************************************
exports.logResult = function(battle, callback){
    var roshamuser = require('d4-roshamuser');
    
    if(!battle){
        return callback({'error': 'What battle you talkin bout?'}, undefined);
    };
    if (!battle.turnDate){
        return callback({'error': 'When was this epic battle of yours?'}, undefined);
    };
    if ((!battle.player1)||(!battle.player2)||(!battle.weapon1)||(!battle.weapon2)){
        return callback({'error': 'This battle is Borked!'}, undefined);
    };
    
    battle.winner = whoWins(battle.weapon1, battle.weapon2);
      
    modResult.create(battle, function(err, newResult){
        if(err){
            return callback({'error': err}, undefined);
        };
        var result1,
            result2,
            resultCount = 0;
        switch(newResult.winner){
            case 0:
                result1 = 'Tie';
                result2 = 'Tie';
                break;
            case 1:
                result1 = 'Win';
                result2 = 'Loss';                
                break;
            case 2:
                result1 = 'Loss';
                result2 = 'Win';
                break;
        }
        var returnResult = function(err, result){
            if (result){
                resultCount++;
                if (resultCount === 2)
                    return callback(undefined, newResult);   
            }
            else
                return callback({'error': 'No can log battle: ' + err.error}, undefined);
        }
        roshamuser.logBattle({'userid':newResult.player1, 'result':result1}, returnResult);
        roshamuser.logBattle({'userid':newResult.player2, 'result':result2}, returnResult);
    });    
}
