// d4-realmuserturn/d4-realmuserturn.js
//******************************************************************************
//  d4-realmuserturn
//  Zot
//  The purpose of this module is to store the user settings for turns through history
//******************************************************************************

var mongoose = require('mongoose');
var conn = require('d4-realmconnection');
var node = require('d4-realmnode');

var weaponTypes = ['Rock','Scissors','Paper'];
var actionTypes = ['Attack', 'Connect', 'Sever'];
var schema = new mongoose.Schema({
    turnDate: { type: Date, default: Date.now },
    user: {type: mongoose.Schema.Types.ObjectId, ref: 'user'}, 
    
    nodes: [{nodeid: {type: mongoose.Schema.Types.ObjectId, ref: 'node'}, 
             weapon: {type: String, enum: weaponTypes}}],

    actions: [{actiontype: {type: String, enum:actionTypes},
               target: mongoose.Schema.Types.ObjectId}]
    
});
var model = mongoose.model('realmUserTurn', schema);
exports.schema = schema;
exports.model = model;
