// d4-realmexplorelog/d4-realmexplorelog.js
var mongoose = require('mongoose');
var schema = new mongoose.Schema({
    createdDate: { type: Date, default: Date.now },
    userID: {type: mongoose.Schema.Types.ObjectId, ref: 'user'}, 
    connUsed: {},
    destNode: Number
});
var model = mongoose.model('explorelog', schema);
exports.schema = schema;
exports.model = model;

//************************************************************************************************************
// function     : d4-realmnexploreLog.addLog()
// developer    : Rob
// developed on : 3/16/2015
// arguments    : {userID, connUsed{}, nodeUsed{}}
//************************************************************************************************************
exports.addLog = function(log, callback){
    if (log.userID && log.connUsed && log.destNode){
        model.create(log, function(err, data){
           if (err)
               callback ({error:err}, undefined);
            else
                callback (undefined, data);
        });
    } else 
        callback ({error: 'Please supply {userID, connUsed{}, nodeUsed{}}'}, undefined);
};

//************************************************************************************************************
// function     : d4-realmnexploreLog.getLogbyUser()
// developer    : Rob
// developed on : 3/16/2015
// arguments    : {userID}
//************************************************************************************************************
exports.getLogbyUser = function(user, callback){
    if (user.userID){
        model.find({userID: user.userID})
            .limit(user.limit)
            .sort(user.sort)
            .exec(function(err, data){
           if (err)
               callback ({error:err}, undefined);
            else
                callback (undefined, data);
        });
    } else 
        callback ({error: 'Please supply {userID}'}, undefined);
};
