// d4-user/d4-user.js
var mongoose = require('mongoose');
var db = mongoose.connection;
var userSchema = new mongoose.Schema({
    email: String,
    name: String,
    password: String
});
var userModel = mongoose.model('user', userSchema);


exports.login = function(authDetails){
}

exports.logout = function(session){
}

exports.register = function(newUser, callback){
    if(db){
        if(newUser){
            console.log(newUser);
            var oUser = {name: newUser.displayName, email: newUser.authentication.details.emailAddress, password: newUser.authentication.details.password}
            //var mUser = new userModel(oUser);
            userModel.create(oUser, function(err, noob){
                if (err) callback({'error': err}, undefined);
                else{
                    callback(undefined, noob);
                }
            });
            
        }
        else{
            console.log('d4-user: error');
            callback({'error': 'naughty, naughty, please provide a user'}, undefined);
        }
    }
    else{
        console.log('d4-user: no database');
        callback({'error': 'd4-user not initialized'}, undefined);
    }
}

exports.update = function(user){
}
