// d4-user/d4-user.js
var mongoose = require('mongoose');
var db = mongoose.connection;
var userSchema = new mongoose.Schema({
    email: String,
    name: String,
    password: String
});
var userModel = mongoose.model('user', userSchema);


exports.login = function(authDetails){
}

exports.logout = function(session){
}


//************************************************************************************************************
// function : d4-user.register
// developer: Kai
// developed on : 5/17/2014
// arguments : newUser object {displayname, authentication.emailaddress, authentication.password}
//             callback function (error, user._id)
//************************************************************************************************************
exports.register = function(newUser, callback){
    if(newUser){
        console.log(newUser);
        var oUser = {
            name: newUser.displayName, 
            email: newUser.authentication.details.emailAddress, 
            password: newUser.authentication.details.password
        }
        var query = userModel.findOne({'email': newUser.authentication.details.emailAddress});
        
        if(newUser.displayName == ''){
            callback({'error': 'That user must have a display name'}, undefined);
            return;
        }
        if(newUser.authentication.details.emailAddress == ''){
            callback({'error': 'That user must have an email address'}, undefined);
            return;
        }
        if(newUser.authentication.details.password == ''){
            callback({'error': 'That user must have a password'}, undefined);
            return;
        }
        
        query.exec(function(err, foundUser){
            if(err){
                callback({'error': err}, undefined);
                return;
            }
            if(!foundUser){
                userModel.create(oUser, function(err, noob){
                    if (err) callback({'error': err}, undefined);
                    else{
                        callback(undefined, noob._id);
                    }
                });
            }
            else{
                callback({'error': 'That user already exists, could not register'}, undefined);
            }
        });
            
    }
    else{
        callback({'error': 'naughty, naughty, please provide a user'}, undefined);
    }
}

exports.update = function(user){
}
