// d4-realmconnection/d4-realmnconnection.js
var mongoose = require('mongoose');
var schema = new mongoose.Schema({
    createdDate: { type: Date, default: Date.now },
    node1: {type: mongoose.Schema.Types.ObjectId, ref: 'node'}, 
    node2: {type: mongoose.Schema.Types.ObjectId, ref: 'node'}, 
    type: {}
});
var model = mongoose.model('connection', schema);
exports.schema = schema;
exports.model = model;

//************************************************************************************************************
// function     : d4-realmnconnection.addConnection()
// developer    : Rob
// developed on : 8/3/2014
// arguments    : {node1, node2}
//************************************************************************************************************
exports.addConnection = function(conn, callback){
    if (conn.node1 && conn.node2 && (conn.node1 != conn.node2)){
        model.create(conn, function(err, data){
           if (err)
               callback ({error:err}, undefined);
            else
                callback (undefined, data);
        });
    } else 
        callback ({error: 'Two unique nodes must be defined for a connection'}, undefined);
};
