// d4-realmconnection/d4-realmnconnection.js
var mongoose = require('mongoose');
var schema = new mongoose.Schema({
    createdDate: { type: Date, default: Date.now },
    node1: {type: mongoose.Schema.Types.ObjectId, ref: 'node'}, 
    desc12: String,
    node2: {type: mongoose.Schema.Types.ObjectId, ref: 'node'}, 
    desc21: String,
    type: {/*tobedefined*/}
});
var model = mongoose.model('connection', schema);
exports.schema = schema;
exports.model = model;

//************************************************************************************************************
// function     : d4-realmnconnection.addConnection()
// developer    : Rob
// developed on : 8/3/2014
// arguments    : {node1, node2}
//************************************************************************************************************
exports.addConnection = function(conn, callback){
    if (conn.node1 && conn.node2 && (conn.node1 != conn.node2)){
        model.create(conn, function(err, data){
           if (err)
               callback ({error:err}, undefined);
            else
                callback (undefined, data);
        });
    } else 
        callback ({error: 'Two unique nodes must be defined for a connection'}, undefined);
};

//************************************************************************************************************
// function     : d4-realmnconnection.updateConnection()
// developer    : Rob
// developed on : 8/3/2014
// arguments    : {connID, desc12, desc21}
//************************************************************************************************************
exports.updateConnection = function(conn, callback){
    if (conn.connID){
        if (conn.desc12 || conn.desc21){
            var update = {};
            if (conn.desc12)
                update.desc12 = conn.desc12;
            if (conn.desc21)
                update.desc21 = conn.desc21;
            
            model.findOneAndUpdate({_id:conn.connID}, update, function(err, data){
               if (err)
                   callback ({error:err}, undefined);
                else
                    callback (undefined, data);
            });
        } else 
            callback ({error: 'There is nothing to update'}, undefined);
    } else
        callback ({error: 'I need a connection dummy'}, undefined);
};

//************************************************************************************************************
// function     : d4-realmnconnection.getConnections()
// developer    : Rob
// developed on : 8/3/2014
// arguments    : {nodes:[nodeIDs]}
//************************************************************************************************************
exports.getConnections = function(conn, callback){
    var query = model.find();
    query.populate('node1');
    query.populate('node2');
    query.or([{node1:{$in:conn.nodes}}, {node2:{$in:conn.nodes}}]);
    query.exec(function(err, data){
       if (err)
           callback ({error:err}, undefined);
        else
            callback (undefined, data);
    });
}
