// d4-roshamleader/d4-roshamleader.js
var mongoose = require('mongoose');
var schema = new mongoose.Schema({
    effectiveDate: Date,
    user: {type: mongoose.Schema.Types.ObjectId, ref: 'user'}, 
    score:  {
        winRate: Number,
        totalWins: Number
    }
});
var model = mongoose.model('leaderboard', schema);
exports.schema = schema;
exports.model = model;

//************************************************************************************************************
// function     : d4-roshamleader.genScores
// developer    : Kai
// developed on : 7/17/2014
// arguments    : 
//************************************************************************************************************
exports.genScores = function(){
    var roshamuser = require('d4-roshamuser');
    var scoreDate = new Date;
    var newScore;
    
    var query = roshamuser.roshamuserModel.find();
    
    query.find(function(err, foundUsers){
        if(foundUsers){
            for(var i=0; i<foundUsers.length; i++){
                newScore = new model({'effectiveDate': scoreDate,
                                      'user': foundUsers[i].userid});
                newScore.score.winRate = foundUsers[i].totalWins / foundUsers[i].totalBattles;
                newScore.score.totalWins = foundUsers[i].totalWins;
                newScore.save(function (err){});
            }
        }
    });
};

//************************************************************************************************************
// function     : d4-roshamleader.getScores
// developer    : Kai
// developed on : 7/17/2014
// arguments    : scoreCard { scoreType: winRate or totalWins
//                            numScores: int
//                            startAtRank: int  }
//************************************************************************************************************
exports.getScores = function(scoreCard, callback){
    callback(undefined, [{'effectiveDate': new Date,
                          'score':{'winRate':'50',
                                   'totalWins':'10'},
                          'user':{'_id':'4321', 'name': 'Zot the Magnificent'}
                         },
                         {'effectiveDate': new Date,
                          'score':{'winRate':'40',
                                   'totalWins':'5'},
                          'user':{'_id':'1234', 'name': 'Vagabond'}
                         }]);
};
