// d4-roshamleader/d4-roshamleader.js
var mongoose = require('mongoose');
var newSchema = new mongoose.Schema({
    effectiveDate: Date,
    userid: {type: mongoose.Schema.Types.ObjectId, ref: 'user'}, 
    score:  {
        winRate: Number,
        totalWins: Number
    }
});
var newMod = mongoose.model('leaderboard', newSchema);
exports.schema = newSchema;
exports.model = newMod;

//************************************************************************************************************
// function     : d4-roshamleader.genScores
// developer    : Kai
// developed on : 7/17/2014
// arguments    : 
//************************************************************************************************************
exports.genScores = function(){
     
    
};

//************************************************************************************************************
// function     : d4-roshamleader.getScores
// developer    : Kai
// developed on : 7/17/2014
// arguments    : 
//************************************************************************************************************
exports.getScores = function(callback){
    callback(undefined, [{'effectiveDate': new Date(),
                          'score':{'winRate':'50',
                                   'totalWins':'10'}
                          'userid':{'_id':'4321', 'name': 'Zot the Magnificent'}
                         },
                         {'effectiveDate': new Date(),
                          'score':{'winRate':'40',
                                   'totalWins':'5'}
                          'userid':{'_id':'1234', 'name': 'Vagabond'}
                         }]);
};
